modules/l4dns/matcher.go

```diff
@@ -0,0 +1,408 @@
+// Copyright 2024 VNXME
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package l4dns
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+	"regexp"
+	"strings"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/miekg/dns"
+
+	"github.com/mholt/caddy-l4/layer4"
+)
+
+func init() {
+	caddy.RegisterModule(&MatchDNS{})
+}
+
+// MatchDNS is able to match connections that look like DNS protocol.
```

Comment:
Should probably document somewhere that if proxying DNS connections, TCP/UDP can't be mixed from inbound to outbound (but to file an issue if this is needed).

Reply:
OK, done.